src/sys/freebsd.rs

```diff
@@ -1 +1,264 @@
+}
+
+pub struct XAttrs {
+    user_attrs: Box<[u8]>,
+    system_attrs: Box<[u8]>,
+    offset: usize,
+}
+
+impl Clone for XAttrs {
+    fn clone(&self) -> Self {
+        XAttrs {
+            user_attrs: Vec::from(&*self.user_attrs).into_boxed_slice(),
+            system_attrs: Vec::from(&*self.system_attrs).into_boxed_slice(),
+            offset: self.offset
+        }
+    }
+
+    fn clone_from(&mut self, other: &XAttrs) {
+        self.offset = other.offset;
+
+        let mut data = mem::replace(&mut self.user_attrs, Box::new([])).into_vec();
+        data.extend(other.user_attrs.iter().cloned());
+        self.user_attrs = data.into_boxed_slice();
+
+        data = mem::replace(&mut self.system_attrs, Box::new([])).into_vec();
+        data.extend(other.system_attrs.iter().cloned());
+        self.system_attrs = data.into_boxed_slice();
+    }
+}
+
+impl Iterator for XAttrs {
+    type Item = OsString;
+    fn next(&mut self) -> Option<OsString> {
+        if self.user_attrs.is_empty() && self.system_attrs.is_empty() {
+            return None;
+        }
+
+        if self.offset == self.user_attrs.len() + self.system_attrs.len() {
+            return None;
+        }
+
+        let data = if self.offset < self.system_attrs.len() {
+            &self.system_attrs[self.offset..]
+        } else {
+            &self.user_attrs[self.offset - self.system_attrs.len() ..]
+        };
+
+        let siz = data[0] as usize;
+
+        self.offset += siz + 1;
+        if self.offset < self.system_attrs.len() {
+            Some(prefix_namespace(OsStr::from_bytes(&data[1..siz + 1]).to_owned(), EXTATTR_NAMESPACE_SYSTEM))
+        } else {
+            Some(prefix_namespace(OsStr::from_bytes(&data[1..siz + 1]).to_owned(), EXTATTR_NAMESPACE_USER))
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        if self.user_attrs.len() + self.system_attrs.len() == self.offset {
+            (0, Some(0))
+        } else {
+            (1, None)
+        }
+    }
+}
+
+fn name_to_ns(name: &OsStr) -> io::Result<(c_int, CString)> {
+    let name_str = match name.to_str() {
+        Some(s) => s,
+        None => return Err(io::Error::new(io::ErrorKind::InvalidInput, "cannot convert to rust string"))
+    };
+
+    let dot = match name_str.find('.') {
+        Some(i) => i,
+        None => return Err(io::Error::new(io::ErrorKind::InvalidInput, "no namespace present in attribute name"))
+    };
+
+    let (nsname, propname) = name_str.split_at(dot);
+    let ns_int = match EXTATTR_NAMESPACE_NAMES.iter().position(|&s| s == nsname) {
+        Some(i) => i,
+        None => return Err(io::Error::new(io::ErrorKind::InvalidInput, "no matching namespace"))
+    };
+
+    Ok((ns_int as c_int, try!(name_to_c(OsStr::new(&propname[1..])))))
+}
+
+fn prefix_namespace(attr: OsString, ns: c_int) -> OsString {
+    let nsname = EXTATTR_NAMESPACE_NAMES[ns as usize];
+    let mut ret = OsString::from(nsname);
```

Comment:
To avoid extra allocations, you can use `Vec::with_capacity(nsname.len() + attr.len() + 1)` to allocate a vec, push the components onto the vec, and then use the unix specific `OsStringExt` trait to convert the `Vec<u8>` into an `OsString`. This isn't *that* important but it's a useful trick to know.